charmcraft/application/main.py

```diff
@@ -42,15 +47,24 @@
 class Charmcraft(Application):
     """Charmcraft application definition."""
 
+    def __init__(
+        self,
+        app: AppMetadata,
+        services: CharmcraftServiceFactory,
+    ) -> None:
+        super().__init__(app=app, services=services)
+        self._global_args: dict[str, Any] = {}
+
     @property
     def command_groups(self) -> list[craft_cli.CommandGroup]:
         """Return command groups."""
-        lifecycle_commands = get_lifecycle_command_group()
+        lifecycle_commands = commands.get_lifecycle_command_group()
```

Comment:
for craft-application 1.1 we could see about maybe adding a `get_lifecycle_command_group()` method that this class can then override

Reply:
Something like that, yeah. Another option would be to provide the full set of command groups to `__init__` with a fallback to generating the default groups.

Reply:
https://github.com/canonical/craft-application/issues/124